hash_table.h

```diff
@@ -1,57 +1,95 @@
 #ifndef HASH_TABLE_H
 #define HASH_TABLE_H
 
-#include <type_traits>  // std::is_pointer
-
+#include <assert.h>
+#include <cstring>
+#include <functional>
+#include <type_traits>
 #include "alloc.h"
-#include "buf.h"
 
-// Parallel buffers of ids (const char *) and value (e.g. Method)
-// for faster searching in the ids. The ids are interned
-// so simple pointer comparison works.
+template <typename T, MEM mem, typename H = std::hash<char *>>
+struct HashTable
+{
+    static_assert(std::is_pointer_v<T> && !std::is_same_v<T, void*>,
+        "T must be a pointer and cannot be void *");
 
-// TODO: T is Q*. Impose that Q should have an `id` field so that we
-// pass one thing on insert()?
+    HashTable() = delete;
 
-// Usage notes:
-// 1) A pointer should always be passed as T.
-// 2) It's the user's responsibility to not insert duplicate `id`.
+    explicit HashTable(size_t buckets)
+        : size{buckets}
+        , ids{static_cast<const char **>(allocate_zero(buckets, mem))}
+        , data{static_cast<T *>(allocate_zero(buckets, mem))}
```

Comment:
For a linear probing (and generally, for a hash table) to be performant, we need the buckets to be quite more than the actual number of elements that will be stored. That can be e.g. `num_buckets = num_elements / 0.7`.
So, there are 2 cases here:
1) `buckets` is the number of elements, in which case it is bad as the hash table will be 100% full and thus will result in bad performance.
2) `buckets` is the number of buckets, in which case it is required that the user computes the number of buckets, which is bad, because that should be esoteric to the table.

Reply:
Also, a `sizeof` is missing that unfortunately, is one of the worst errors to debug.